hilti/toolchain/src/compiler/visitors/constant-folder.cc

```diff
@@ -0,0 +1,357 @@
+        if ( auto ctor = foldConstant<Ctor>(callArgument(op, 0)) ) {
+            auto x = cb(*ctor);
+            x.setMeta(p.node.meta());
+            return {std::move(x)};
+        }
+        else
+            return std::nullopt;
+    }
+
+    // Helper to extract the 1st argument of a call expression.
+    Expression callArgument(const expression::ResolvedOperatorBase& o, int i) {
+        auto ctor = o.op1().as<expression::Ctor>().ctor();
+
+        if ( auto x = ctor.tryAs<ctor::Coerced>() )
+            ctor = x->coercedCtor();
+
+        return ctor.as<ctor::Tuple>().value()[i];
+    }
+
+    // Helper to cast an uint64 to int64, with range check.
+    int64_t to_int64(uint64_t x, position_t& p) {
+        try {
+            return SafeInt<int64_t>(x);
+        } catch ( ... ) {
+            throw hilti::rt::OutOfRange("integer value out of range");
+        }
+    }
+
+    // Helper to cast an int64 to uint64, with range check.
+    uint64_t to_uint64(int64_t x, position_t& p) {
+        if ( x < 0 )
+            throw hilti::rt::OutOfRange("integer value out of range");
+
+        return static_cast<uint64_t>(x);
+    }
+
+    result_t operator()(const expression::Ctor& n, position_t p) { return n.ctor(); }
+
+    result_t operator()(const operator_::signed_integer::SignNeg& n, position_t p) {
+        auto op = foldConstant<ctor::SignedInteger>(n.op0());
+        if ( ! op )
+            return std::nullopt;
+
+        return ctor::SignedInteger(-op->value(), op->width(), p.node.meta());
+    }
+
+    result_t operator()(const expression::Grouping& n, position_t p) {
+        auto x = _foldConstant(n.expression());
+        if ( ! x )
+            return std::nullopt;
+
+        return *x;
+    }
+
+    result_t operator()(const expression::LogicalOr& n, position_t p) {
+        auto op0 = foldConstant<ctor::Bool>(n.op0());
+        auto op1 = foldConstant<ctor::Bool>(n.op1());
+        if ( ! (op0 && op1) )
+            return std::nullopt;
+
+        return ctor::Bool(op0->value() || op1->value(), p.node.meta());
+    }
+
+    result_t operator()(const expression::LogicalAnd& n, position_t p) {
+        auto op0 = foldConstant<ctor::Bool>(n.op0());
+        auto op1 = foldConstant<ctor::Bool>(n.op1());
+        if ( ! (op0 && op1) )
+            return std::nullopt;
+
+        return ctor::Bool(op0->value() && op1->value(), p.node.meta());
+    }
+
+    result_t operator()(const expression::LogicalNot& n, position_t p) {
+        auto op = foldConstant<ctor::Bool>(n.expression());
+        if ( ! op )
+            return std::nullopt;
+
+        return ctor::Bool(! op->value(), p.node.meta());
+    }
+
+    result_t operator()(const expression::ResolvedID& n, position_t p) {
+        // We cannot fold the optimizer's feature constants currently because
+        // that would mess up its state tracking. We continue to let the
+        // optimizer handle expressions involving these.
+        //
+        // TODO: Can we unify this?
```

Comment:
The issue here is that feature constants are not really constants, but instead flags we toggle in the optimizer if we can detect that they are unused. If we would move this detection before this folder we could treat them like any other constant, but that seems tricky.